File: epi_judge_java_solutions/epi/BstFromPreorder.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;

public class BstFromPreorder {
  @EpiTest(testDataFile = "bst_from_preorder.tsv")

  public static BstNode<Integer>
  rebuildBSTFromPreorder(List<Integer> preorderSequence) {

    return rebuildBSTFromPreorderHelper(preorderSequence, 0,
                                        preorderSequence.size());
  }

  // Builds a BST from preorderSequence.subList(start, end).
  private static BstNode<Integer>
  rebuildBSTFromPreorderHelper(List<Integer> preorderSequence, int start,
                               int end) {
    if (start >= end) {
      return null;
    }
    int transitionPoint = start + 1;
    while (transitionPoint < end &&
           Integer.compare(preorderSequence.get(transitionPoint),
                           preorderSequence.get(start)) < 0) {
      ++transitionPoint;
    }
    return new BstNode<>(
        preorderSequence.get(start),
        rebuildBSTFromPreorderHelper(preorderSequence, start + 1,
                                     transitionPoint),
        rebuildBSTFromPreorderHelper(preorderSequence, transitionPoint, end));
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "BstFromPreorder.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp_solutions/bst_from_preorder.cc
#include <algorithm>
#include <memory>

#include "bst_node.h"
#include "test_framework/generic_test.h"

using std::find_if_not;
using std::make_unique;
using std::unique_ptr;
using std::vector;

unique_ptr<BstNode<int>> RebuildBSTFromPreorderHelper(const vector<int>&, int,
                                                      int);

unique_ptr<BstNode<int>> RebuildBSTFromPreorder(
    const vector<int>& preorder_sequence) {
  return RebuildBSTFromPreorderHelper(preorder_sequence, 0,
                                      size(preorder_sequence));
}

// Builds a BST from preorder_sequence[start, end - 1].
unique_ptr<BstNode<int>> RebuildBSTFromPreorderHelper(
    const vector<int>& preorder_sequence, int start, int end) {
  if (start >= end) {
    return nullptr;
  }

  int transition_point = distance(
      cbegin(preorder_sequence),
      find_if_not(cbegin(preorder_sequence) + start, cend(preorder_sequence),
                  [&](int a) { return a <= preorder_sequence[start]; }));
  return make_unique<BstNode<int>>(
      preorder_sequence[start],
      RebuildBSTFromPreorderHelper(preorder_sequence, start + 1,
                                   transition_point),
      RebuildBSTFromPreorderHelper(preorder_sequence, transition_point, end));
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"preorder_sequence"};
  return GenericTestMain(args, "bst_from_preorder.cc", "bst_from_preorder.tsv",
                         &RebuildBSTFromPreorder, DefaultComparator{},
                         param_names);
}
